src/clr-angular/forms/radio/radio-container.ts

```diff
@@ -4,23 +4,67 @@
  * The full license information can be found in LICENSE in the root directory of this project.
  */
 
-import { Component } from '@angular/core';
+import { Component, ContentChild, OnDestroy, Optional } from '@angular/core';
+import { Subscription } from 'rxjs';
 
 import { DynamicWrapper } from '../../utils/host-wrapping/dynamic-wrapper';
-import { ControlIdService } from '../common/providers/control-id.service';
+import { IfErrorService } from '../common/if-error/if-error.service';
+import { ClrLabel } from '../common/label';
+import { ControlClassService } from '../common/providers/control-class.service';
+import { LayoutService } from '../common/providers/layout.service';
+import { NgControlService } from '../common/providers/ng-control.service';
 
 @Component({
   selector: 'clr-radio-container',
   template: `
-        <!-- We want the radio input to be before the label, always -->
-        <ng-content select="[clrRadio]"></ng-content>
-        <ng-content></ng-content>
-        <label *ngIf="_dynamic"></label>
+    <ng-content select="label"></ng-content>
+    <label *ngIf="!label && addGrid()"></label>
+    <div class="clr-control-container" [ngClass]="controlClass()">
+      <ng-content select="clr-radio-wrapper"></ng-content>
+      <ng-content select="clr-control-helper" *ngIf="!invalid"></ng-content>
+      <clr-icon *ngIf="invalid" class="clr-validate-icon" shape="exclamation-circle"></clr-icon>
+      <ng-content select="clr-control-error" *ngIf="invalid"></ng-content>
```

Comment:
So the error is per-radio. I know we keep having that discussion, but I really don't remember us going that way for radio buttons. It makes sense for checkboxes, but radios are a single input so it validates as one.

Reply:
The error is per radio based on the form control, so if you have 3 radios by the same name they share the same error state. I think you have it backwards, having 3 radios each with `name="control"` will show only one validation error, which is most cases I can imagine is only that it is required. Some custom validation might be created where certain radios are invalid given other conditions, but either way the control for all radios inside of the container should be marked as `invalid` and therefore the error messages should appear.

I capture the last declared radio for their validity, because Angular updates all radio controllers with the same validity state and I only need one of them to do the checks. Checkboxes are a different beast, which is partially why I started with radios.

Reply:
Oh I misread this, my bad.

👍